### ISSUE TITLE
add nn.CenterCrop1d and nn.CenterCrop2d

### ISSUE THREAD

**Issue by bodokaiser** [pytorch/pytorch#1331]
There are some models which use "crop layers" (e.g. [U-Net](https://arxiv.org/abs/1505.04597), [ParseNet](https://arxiv.org/abs/1506.04579)) so I think it wouldn't be bad to have a `CenterCropNd` layer.

As `F.pad` supports negative padding we just need to calculate padding offsets on top.

```python
import torch

from torch.nn import functional as F
from torch.autograd import Variable

def center_crop(x, height, width):
    crop_h = torch.FloatTensor([x.size()[2]]).sub(height).div(-2)
    crop_w = torch.FloatTensor([x.size()[3]]).sub(width).div(-2)

    return F.pad(x, [
        crop_w.ceil().int()[0], crop_w.floor().int()[0],
        crop_h.ceil().int()[0], crop_h.floor().int()[0],
    ])

variable = Variable(torch.randn(1, 3, 60, 40))

print(center_crop(variable, 20, 20).size())
print(center_crop(variable, 20, 40).size())
```

I can send a PR on interest.

**Comment by fmassa** [comment]
Hum, I think the type of cropping that one might want to apply might be problem-specific (not necessarily only center crop).
Plus, cropping an image can be performed as efficiently as `F.pad` by indexing the tensor with slices like `x[:, :, 3:7, 5:10]`, so I'm not so sure about the need of a special layer for that.

**Comment by bodokaiser** [comment]
What about [caffe's crop layer](http://caffe.berkeleyvision.org/tutorial/layers/crop.html)?

**Comment by fmassa** [comment]
Caffe doesn't give you the ability to easily manipulate the tensors, so there is a need of such layer. The same applies for lua torch, as you need to define a module that will compute the operation that can be differentiated, but in pytorch you can just directly manipulate the tensors and the backprop will be performed.
if you check caffe's crop layer, it's almost the same as the snippet I sent in the previous message.

**Comment by apaszke** [comment]
Even though it's easy to do it with indexing, I think it'd be better to provide some standard crops in the library. It's much more convenient then messing with indices, and can save making a few errors along the way each time you need to crop.

**Comment by andrewgiessel** [comment]
+1 to this.  I think standard cropping layers would be great!

**Comment by bodokaiser** [comment]
I could make a PR this week if there is consensus on API?

**Comment by apaszke** [comment]
I think `F.center_crop(x, *args)` and `nn.CenterCrop(*args)` (where `len(args) == x.dim() - 2`) would be the best API for it.

**Comment by andrewgiessel** [comment]
Is it reasonable to include non-symmetric cropping in this API as well?

**Comment by bodokaiser** [comment]
@andrewgiessel See the current proposition at #1349. I think it would be also reasonable to specify center crop in more detail (e.g. you want to crop 10 -> 3 you have one offset 3 and one 4 but you cants specify the order at the moment).
Is this what you thought of?

**Comment by fmassa** [comment]
This can be achieved by passing a negative pad in `torch.nn.functional.pad`.
